File: app/src/main/java/com/qimian233/ztool/hook/modules/launcher/grid/CustomGridSize.java
package com.qimian233.ztool.hook.modules.launcher.grid;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;

import com.qimian233.ztool.hook.base.AppHookModule;

import io.github.libxposed.api.XposedModuleInterface;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;

@SuppressLint("PrivateApi")
public class CustomGridSize extends AppHookModule {
    private static int CUSTOM_COLUMNS = 8;
    private static int CUSTOM_ROWS = 6;

    public CustomGridSize() {}

    @Override
    public String getModuleName() {
        return "CustomGridSize";
    }
    @Override
    public String[] getTargetPackages() {
        return new String[]{"com.zui.launcher"};
    }
    @Override
    public void handleLoadPackage(XposedModuleInterface.PackageLoadedParam param) throws Throwable {
        ClassLoader classLoader = param.getDefaultClassLoader();
        logger.info("Load CustomGridSize!");
        // We directly hook the constructor of GridOption class
        // But before hook, let us load custom grid size from shared prefs first
        getCustomGridSize();
        try {
            // First find GridOption class (I DO NOT believe Lenovo will mod this class)
            final Class<?> gridOptionClass;
            try {
                gridOptionClass = classLoader.loadClass("com.android.launcher3.InvariantDeviceProfile$GridOption");
            } catch (ClassNotFoundException e) {
                logger.error("GridOption class not found on this ROM");
                return;
            }
            logger.debug("Found GridOption class!");
            // Then formally start our job
            // Find arg class to construct correct method signature
            Class<?> contextClass = Context.class;
            Class<?> attributeSetClass = classLoader.loadClass("android.util.AttributeSet");
            Class<?> displayInfoClass = null;
            try {
                displayInfoClass = classLoader.loadClass("com.android.launcher3.util.DisplayController$Info");
            } catch (ClassNotFoundException ignored) {
                logger.error("Unable to find DisplayInfo class");
            }
            Constructor<?> ctor;
            if (displayInfoClass != null) {
                try {
                    ctor = gridOptionClass.getDeclaredConstructor(
                            contextClass, attributeSetClass, displayInfoClass);
                } catch (Exception e) {
                    logger.error("Exception happened when trying to find GridOption class with constructor signature Context, AttributeSet, DisplayController$Info: ", e);
                    try {
                        ctor = gridOptionClass.getDeclaredConstructor(
                                contextClass, attributeSetClass);
                    } catch (Exception ignored) {
                        logger.error("Failed to get constructor with alternate way, exiting.");
                        return;
                    }
                }
            } else {
                logger.warn("Cannot find DisplayController$Info, use alternate constructor signature.");
                try {
                    ctor = gridOptionClass.getDeclaredConstructor(
                            contextClass, attributeSetClass);
                }
                catch (Exception e) {
                    logger.error("Failed to find constructor, exiting.", e);
                    return;
                }
            }
            hookWithId(ctor, "ctor", chain -> {
                chain.proceed();
                try {
                    Object thisObject = chain.getThisObject();
                    // Directly set int fields
                    Field numColsField = gridOptionClass.getDeclaredField("numColumns");
                    numColsField.setAccessible(true);
                    numColsField.set(thisObject, CUSTOM_COLUMNS);

                    Field numRowsField = gridOptionClass.getDeclaredField("numRows");
                    numRowsField.setAccessible(true);
                    numRowsField.set(thisObject, CUSTOM_ROWS);

                    logger.debug("GridOption config modded to " + CUSTOM_COLUMNS + "x" + CUSTOM_ROWS);
                } catch (Exception e) {
                    logger.error("No such method! Probably you are using a newer ZUXOS version!", e);
                }
                return null;
            });
        } catch (Exception e) {
            logger.error("Failed to hook GridOption!", e);
        }
    }

    private void getCustomGridSize() {
        SharedPreferences prefs = this.xposed.getRemotePreferences("xposed_module_config");
        CUSTOM_ROWS = prefs.getInt("CustomLauncherRow", 4);
        CUSTOM_COLUMNS = prefs.getInt("CustomLauncherColumn", 6);
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/launcher/grid/CustomGridSize.kt
package com.qimian233.ztool.hook.modules.launcher.grid

import android.annotation.SuppressLint
import android.content.Context
import com.qimian233.ztool.data.PreferenceKeys
import com.qimian233.ztool.hook.base.AppHookModule
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam
import java.lang.reflect.Constructor

@SuppressLint("PrivateApi")
class CustomGridSize : AppHookModule() {
    override fun getModuleName(): String = "CustomGridSize"

    override fun getTargetPackages(): Array<String> = arrayOf("com.zui.launcher")

    @Throws(Throwable::class)
    override fun handleLoadPackage(param: PackageLoadedParam) {
        val classLoader = param.defaultClassLoader
        logger.info("Load CustomGridSize!")
        // We directly hook the constructor of GridOption class
        // But before hook, let us load custom grid size from shared prefs first
        val prefs = remotePreferences
        CUSTOM_ROWS =
            prefs.getInt(PreferenceKeys.CUSTOM_LAUNCHER_ROW.name, 4)
        CUSTOM_COLUMNS =
            prefs.getInt(PreferenceKeys.CUSTOM_LAUNCHER_COLUMN.name, 6)
        try {
            val gridOptionClass: Class<*> = classLoader.loadClass($$"com.android.launcher3.InvariantDeviceProfile$GridOption")
            // Find arg class to construct correct method signature
            val contextClass: Class<*> = Context::class.java
            val attributeSetClass = classLoader.loadClass("android.util.AttributeSet")
            val displayInfoClass: Class<*> = classLoader.loadClass($$"com.android.launcher3.util.DisplayController$Info")
            var ctor: Constructor<*>?
            try {
                ctor = gridOptionClass.getDeclaredConstructor(
                    contextClass, attributeSetClass, displayInfoClass
                )
            } catch (e: Exception) {
                logger.error(
                    $$"Exception happened when trying to find GridOption class with constructor signature Context, AttributeSet, DisplayController$Info: ",
                    e
                )
                try {
                    ctor = gridOptionClass.getDeclaredConstructor(
                        contextClass, attributeSetClass
                    )
                } catch (_: Exception) {
                    logger.error("Failed to get constructor with alternate way, exiting.")
                    return
                }
            }
            hookWithId(ctor, "ctor") { chain ->
                chain.proceed()
                try {
                    val thisObject = chain.thisObject
                    // Directly set int fields
                    val numColsField = findField(gridOptionClass, "numColumns")
                    numColsField.set(thisObject, CUSTOM_COLUMNS)
                    val numRowsField = findField(gridOptionClass, "numRows")
                    numRowsField.set(thisObject, CUSTOM_ROWS)
                    logger.debug("GridOption config modded to " + CUSTOM_COLUMNS + "x" + CUSTOM_ROWS)
                } catch (e: Throwable) {
                    logger.error("No such method! Probably you are using a newer ZUXOS version!", e)
                }
                null
            }
        } catch (th: Throwable) {
            logger.error("Failed to hook GridOption!", th)
        }
    }

    companion object {
        private var CUSTOM_COLUMNS = 8
        private var CUSTOM_ROWS = 6
    }
}
